edb/common/asyncutil.py

```diff
@@ -62,3 +62,78 @@ async def deferred_shield(arg: Awaitable[_T]) -> _T:
     if ex:
         raise ex
     return task.result()
+
+
+async def debounce(
+    input: Callable[[], Awaitable[_T]],
+    output: Callable[[list[_T]], Awaitable[None]],
+    *,
+    max_wait: float,
+    delay_amt: float,
+    max_batch_size: int,
+) -> None:
+    '''Debounce and batch async events.
+
+    Loops forever unless an operation fails, so should probably be run
+    from a task.
+
+    The basic algorithm is that if an event comes in less than
+    `delay_amt` since the previous one, then instead of sending it
+    immediately, we wait an additional `delay_amt` from then. If we are
+    already waiting, any message also extends the wait, up to
+    `max_wait`.
+
+    Also, cap the maximum batch size to `max_batch_size`.
+    '''
+    # I think the algorithm reads more clearly with the params
+    # capitalized as constants, though we don't want them like that in
+    # the argument list, so reassign them.
+    MAX_WAIT, DELAY_AMT, MAX_BATCH_SIZE = max_wait, delay_amt, max_batch_size
+
+    loop = asyncio.get_running_loop()
+
+    batch = []
+    last_signal = -MAX_WAIT
+    target_time = None
+
+    while True:
+        try:
+            async with asyncio.timeout_at(target_time):
+                v = await input()
```

Comment:
Hm, so you will be waiting with timeout even for the first `input()` return, but that's not what we want to do. We don't want to keep looping through timeouts until `input()` returns for the first time.

We want to start debouncing only when it returns. The first await on it shouldn't be with a timeout at all.

Reply:
Oh sigh, this is why I don't like None for cases like this. Please add an `if`.

```
if target_time is not None:
    async with ...
else:
    await input()
```